### ISSUE TITLE
smooth_knn_dist crashes when using Numba with fastmath=True on some computers

### ISSUE THREAD

**Issue by mikeriedel** [lmcinnes/umap#246]
I isolated a crash that is repeatable on my machine. The function smooth_knn_dist and the use of numba with fastmath=true are causing problems. Setting fastmath=false for this function, causes the crash to go away and allows me to complete my simple test case as well as run fit_transform.

The exception I had was : 
`Windows fatal exception: access violation`

My test case is : 
`import numpy as np`
`import umap`
`np_features = np.random.rand(100, 100)`
`sigmas, rhos = umap.umap_.smooth_knn_dist(np_features, 20)`

I tried turning off fastmath (line 47 of umap_.py) and the crash went away.
`@numba.njit(parallel=True, fastmath=False)`.

_Originally posted by @mikeriedel in https://github.com/lmcinnes/umap/issues/243#issuecomment-497528216_

**Comment by lmcinnes** [comment]
Thanks -- that is definitely odd. I'll have to do some benchmarking to determine what the performance impact on turning that off actually is.

**Comment by mikeriedel** [comment]
To answer your questions from the other thread. No there is nothing special about this computer. My colleagues are running almost exactly the same configuration. The only variances may be slight hardware differences. 

Regarding fastmath. My reading of the documentation suggests that using fastmath while working with infinity is undefined. I will play around with this a little on Monday. 

> No Infs - Allow optimizations to assume the arguments and result are not +/-Inf. If an argument is +/-Inf, or the result would be +/-Inf, it produces a poison value instead.
https://llvm.org/docs/LangRef.html#fast-math-flags

**Comment by mikeriedel** [comment]
Looks like it was a Numba bug. I updated Numba to 0.44.0 and the problem went away. 

In reference to my previous comment, I removed the use of Infinity with fastmath in smooth_knn_dist and that did not make a difference.

For archive purposes if anyone else has this problem: My previous version of Numba was 0.43.1.  Before I tried updating Numba, I also tried updating LLVMlite and Numpy which had no affect. Versions: LLVMlite (0.28 -> 0.29) and Numpy (1.15.4 -> 1.16.4)

Thank you for your help!
